### PR TITLE
[MXNET-100] Support float16 in Correlation operator 

### PR DESCRIPTION
**Pull request by haojin2**
## Description ##
Add support for any datatype for Correlation operator, which is not mentioned in issue #2302.

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [x] Code is well-documented: 
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Change Correlation operator from only supporting real_t to supporting any datatype
- [x] Add additional test cases for float16

### CODE REVIEW

File: src/operator/correlation-inl.h [hunk 1]
```diff
@@ -79,14 +79,14 @@ class CorrelationOp : public Operator {
...
-    tmp1 = 0.0f;
-    tmp2 = 0.0f;
-    out = 0.0f;
+    Tensor<xpu, 4, DType> data1 = in_data[Correlation::kData1].get<xpu, 4, DType>(s);
+    Tensor<xpu, 4, DType> data2 = in_data[Correlation::kData2].get<xpu, 4, DType>(s);
+    Tensor<xpu, 4, DType> out   = out_data[Correlation::kOut].get<xpu, 4, DType>(s);
+    Tensor<xpu, 4, DType> tmp1  = out_data[Correlation::kTemp1].get<xpu, 4, DType>(s);
+    Tensor<xpu, 4, DType> tmp2  = out_data[Correlation::kTemp2].get<xpu, 4, DType>(s);
+    tmp1 = DType(0.0f);
+    tmp2 = DType(0.0f);
```
Comment (anirudh2290) [inline src/operator/correlation-inl.h]: Preferably, Use static_cast.

File: src/operator/correlation-inl.h [hunk 2]
```diff
@@ -228,6 +228,22 @@ void Init(const std::vector<std::pair<std::string, std::string> >& kwargs) overr
...
+  bool InferType(std::vector<int> *in_type,
+                 std::vector<int> *out_type,
+                 std::vector<int> *aux_type) const override {
+    int dtype = (*in_type)[0];
+    type_assign(&(*in_type)[1], dtype);
+    type_assign(&(*out_type)[0], dtype);
+    type_assign(&(*out_type)[1], dtype);
+    type_assign(&(*out_type)[2], dtype);
+
+    TYPE_ASSIGN_CHECK(*in_type, 0, dtype);
```
Comment (anirudh2290) [inline src/operator/correlation-inl.h]: Why do you need both type_assign and TYPE_ASSIGN_CHECK ?

Comment (haojin2) [reply src/operator/correlation-inl.h]: It was advice from Jun, he suggested that we can do a mutual inference by reducing all datatypes to one and then assign the reduced type back to everything.

File: src/operator/correlation-inl.h (hunk 1 above)
Comment (haojin2) [reply src/operator/correlation-inl.h]: Sure, will address this issue shortly

File: src/operator/correlation-inl.h (hunk 2 above)
Comment (haojin2) [reply src/operator/correlation-inl.h]: If the reduced datatype is not -1 then all inputs and outputs will then have the same datatype at the end of this function.

File: src/operator/correlation-inl.h (hunk 1 above)
Comment (haojin2) [reply src/operator/correlation-inl.h]: Seems like static_cast is causing some compilation errors, taking a look at it now.

**Comment by piiswrong** [comment]
float16 doesn't work with static_cast

**Pull request merged**